proposals/Layer-1.md

```diff
@@ -0,0 +1,390 @@
+block.
+
+If a throw occurs within a function body, and it doesn't appear inside the body
+of a try block, the throw continues up the call stack until it is in the body of
+an an enclosing try block, or the call stack is flushed. If the call stack is
+flushed, the host VM defines how to handle uncaught exceptions.  Otherwise, the
+found enclosing try block is the catching try block.
+
+A throw inside the body of a catch block is never caught by the corresponding
+try block of the catch block, since instructions in the body of the catch block
+are not in the body of the try block.
+
+Once a catching try block is found for the throw, the operand stack is popped back
+to the size the operand stack had when the try block was entered, and then
+the caught exception is pushed onto the stack.
+
+If control is transferred to the body of a catch block, and the last instruction
+in the body is executed, control then exits the try block.
+
+If the selected catch block does not throw an exception, it must yield the
+value(s) expected by the corresponding catching try block. This includes popping
+the caught exception.
+
+Note that a caught exception can be rethrown using the `throw` instruction.
+
+### Exception data extraction
+
+The `if_except block` defines a conditional query of the exception on top of the
+stack. The exception is not popped when queried. The if_except block has two
+subblocks, the `then` and `else` subblocks, like that of an `if` block. The then
+block is a sequence of instructions following the `if_except` instruction. The
+else block is optional, and if it appears, it begins with the `else`
+instruction.  The scope of the if_except block is from the `if_except`
+instruction to the corresponding `end` instruction.
+
+That is, the forms of an if_except block is:
+
+```
+if_except block_type except_index
+  Instruction*
+end
+
+if_except block_type except_index
+  Instruction*
+else
+  Instructions*
+end
+```
+
+The conditional query of an exception succeeds when the exception on the top of
+the stack is an instance of the corresponding tagged exception type (defined by
+`except_index`).
+
+If the query succeeds, the data values (associated with the type signature of
+the exception class) are extracted and pushed onto the stack, and control
+transfers to the instructions in the then block.
+
+If the query fails, it either enters the else block, or transfer control to the
+end of the if_except block if there is no else block.
+
+### Debugging
+
+Earlier discussion implied that when an exception is thrown, the runtime will
+pop the operand stack across function calls until a corresponding, enclosing try
+block is found. The implementation may actually not do this. Rather, it may
+first search up the call stack to see if there is an enclosing try. If none are
+found, it could terminate the thread at the point of the throw. This would
+allow better debugging capability, since the corresponding call stack is still
+there to query.
```

Comment:
Alternatively, the embedder can capture a stack trace at the point the exception object is created, which I think is roughly what JavaScript will do.

One question is whether stack traces should be captured at the point of creating the exception, or at the point of throwing the exception. I think this is irrelevant to the core spec, since stack traces are provided purely by the embedder and we provide no way to access them in WebAssembly. It's worth discussing in the context of the JS embedding though.

Reply:
Discuss in #37

Reply:
I don't disagree with this. I only wanted to state that it is up to the host on how to best handle this, and, depending on whether it is compiled for debugging or release, do slightly different implementations.

Reply:
Sounds good.

Reply:
If we are going to do the two-phase unwinding, this may have to be incorporated into wasm spec, because the toolchain (more precisely, the library like libcxxabi) has to support this behavior.

For example, for x86 EH, in the first phase, libcxxabi does not actually unwind the stack but just walks up the stack to see if there is a matching handler. If a matching handler is found, in the second phase it actually starts unwinding the stack until it reaches the handler found. If no matching handler is found, libcxxabi does not unwind the stack but just crashes at the point of throwing, leaving the whole stack trace intact.

I'm not suggesting we do this now, and for Layer 1 I think we don't need to do this anyway. I'm just suggesting if we do this later, this can't be just left to the discretion of the host VM but explicitly needs to be specified in the spec.